File: bundle.js
// node bundle.js a.js y.js

import fs from 'fs'

const base64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789$_'

function parse (file, imported, modules) {
  let f, j, k, line, name, names, remove, split, string, text
  try {
    text = fs.readFileSync(file, 'utf8')
  } catch {
    return ''
  }
  while (text.includes(' \n')) {
    text = text.replaceAll(' \n', '\n')
  }
  while (text.includes('\n\n')) {
    text = text.replaceAll('\n\n', '\n')
  }
  let lines = text.split('\n')
  text = ''
  for (line of lines) {
    if (line.trimStart().startsWith('//')) {
      continue
    }
    if (!remove && line.includes('/*') && !line.includes('//*')) {
      if (line.includes('*/')) {
        line = line.slice(0, line.indexOf('/*')) + ' ' + line.slice(line.indexOf('*/') + 2)
      } else {
        line = line.slice(0, line.indexOf('/*'))
        remove = true
      }
    }
    if (remove) {
      if (line.includes('*/')) {
        line = line.slice(line.indexOf('*/') + 2)
        remove = false
      } else {
        continue
      }
    }
    if (line.length && line.replaceAll(' ', '').length) {
      text += line.trimEnd(' ') + '\n'
    }
  }
  const texta = text

  function resolve (f) {
    if (f.slice(0, 2) == './') {
      f = f.slice(2)
    }
    if (f[0] != '.' && f[0] != '/') {
      f = file.split('/').slice(0, -1).join('/') + '/' + f
    } else if (f.startsWith('../')) {
      let i = 0
      while (f.startsWith('../')) {
        f = f.slice(3)
        i += 1
      }
      split = file.split('/').slice(0, -1 - i)
      f = split.length ? split + '/' + f : f
    }
    return f.slice(-3) == '.js' ? f : f + '.js'
  }

  const files = {}
  let i = text.indexOf('import ')
  while (i > -1) {
    if (i != 0 && !['\n', ' '].includes(text[i - 1])) {
      text = text.slice(i + 6)
      i = text.indexOf('import ')
      continue
    }
    text = text.slice(i)
    i = 6
    while (text[i] == ' ') {
      i += 1
    }
    if (text[i] == '{') {
      text = text.slice(i)
      k = text.indexOf('}')
      names = text.slice(1, k).split(',')
      names = names.map(name => name.replaceAll(' ', ''))
      names = names.filter(name => !['', '{', '}'].includes(name))
      text = text.slice(k)
      if (text.includes(' from ')) {
        i = text.indexOf(' from ') + 6
        while (text[i] == ' ') {
          i += 1
        }
        if (text[i] == "'") {
          text = text.slice(i + 1)
          f = resolve(text.slice(0, text.indexOf("'")))
          f in files ? files[f].push(...names) : files[f] = names
        }
      }
    } else {
      names = []
      if (text.includes(' from ')) {
        j = text.indexOf(' ')
        k = text.indexOf(' from ')
        while (j < k) {
          while (text[i] == ' ') {
            i += 1
          }
          name = text.slice(i, k).split(' ')[0]
          if (!['', '{', '}'].includes(name)) {
            names.push(name)
          }
          i = j
          text = text.slice(j)
          if (text.includes(',')) {
            j = text.indexOf(',')
          }
          k = text.indexOf(' from ')
        }
        text = text.slice(k)
        i = 6
      }
      while (text[i] == ' ') {
        i += 1
      }
      if (text[i] == "'") {
        text = text.slice(i + 1)
        f = resolve(text.slice(0, text.indexOf("'")))
        f in files ? files[f].push(...names) : files[f] = names
      }
    }
    i = text.indexOf('import ')
  }
  modules[file] = Object.keys(files)
  if (modules[file].filter(i => !imported.includes(i) && (!(i in modules) || !modules[i].includes(file))).length) {
    return ''
  }
  const exporta = ['async', 'class', 'const', 'default', 'function', 'let', 'var']
  const repeata = ['\n', ' ', '(', ',', '.', '[']
  text = texta
  i = text.indexOf('export ')
  while (i > -1) {
    text = text.slice(i + 7)
    for (name of exporta) {
      j = text.indexOf(name)
      if (j > -1 && j < 3) {
        text = text.slice(j + name.length)
      }
    }
    if (text.includes('\n')) {
      names = text.slice(0, text.indexOf('\n'))
    }
    split = []
    if (!names.includes('=') || names.includes('(') && names.indexOf('=') > names.indexOf('(')) {
      split.push(names)
    } else {
      while (names.includes('=')) {
        split.push(names.slice(0, names.indexOf('=')))
        names = names.slice(names.indexOf('='))
        names = names.includes(',') ? names.slice(names.indexOf(',')) : ''
      }
    }
    names = []
    for (name of split) {
      while (repeata.includes(name[0])) {
        name = name.slice(1)
      }
      for (j of repeata) {
        if (name.includes(j)) {
          name = name.slice(0, name.indexOf(j))
        }
      }
      names.push(name)
    }
    file in files ? files[file].push(...names) : files[file] = names
    i = text.indexOf('export ')
  }

  function replace (text, past, next) {
    const length = past.length
    let a = 0, cont, textb
    while (text.slice(a).includes(past)) {
      a += text.slice(a).indexOf(past)
      if (text.length < a + 1 + length) {
        return text
      }
      cont = false
      textb = text.slice(a - 7, a) + next
      for (name of exporta) {
        if (textb.includes(name + '_')) {
          cont = true
        }
      }
      if (cont || base64.includes(text[a + length]) || (base64 + "'.").includes(text[a - 1])) {
        a += length
        continue
      }
      text = text.slice(0, a) + next + text.slice(a + length)
      a += next.length
    }
    return text
  }

  text = texta
  for (file in files) {
    string = file.replaceAll('.', '_').replaceAll('/', '_')
    split = string.split('_')
    string = string.slice(0, -split[split.length - 1].length - 1)
    for (name of files[file]) {
      text = replace(text, name, name + '_' + string)
    }
  }
  lines = text.split('\n')
  text = ''
  for (line of lines) {
    if (line.startsWith('export default ')) {
      line = line.slice(line.indexOf('export default ') + 15)
    }
    if (line.startsWith('export ')) {
      line = line.slice(line.indexOf('export ') + 7)
    }
    if (line.length && !line.startsWith('import ')) {
      text += line + '\n'
    }
  }
  return text
}

function build (file, output) {
  const imported = []
  let imports = [file]
  const modules = {}
  const texts = {}
  while (imports.length) {
    file = imports[0]
    if (imported.includes(file)) {
      imports = imports.filter(i => i != file)
    } else {
      texts[file] = parse(file, imported, modules)
      imports.unshift(...modules[file])
      if (texts[file].length) {
        imported.push(file)
      }
    }
  }
  let text = ''
  imports = []
  for (file of imported) {
    if (!imports.includes(file)) {
      text += texts[file]
      imports.push(file)
    }
  }
  fs.writeFileSync(output, text)
}

const args = process.argv
build(args[2], args[3])

File: bundle.py
# python3 bundle.py a.js y.js

import sys

base64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789$_'

def parse (file, imported, modules):
    try:
        with open(file, 'r') as text:
            text = text.read()
    except:
        return ''
    while text.find(' \n') > -1:
        text = text.replace(' \n', '\n')
    while text.find('\n\n') > -1:
        text = text.replace('\n\n', '\n')
    lines = text.split('\n')
    text = ''
    remove = False
    for line in lines:
        if line.lstrip().startswith('//'):
            continue
        if not remove and '/*' in line and '//*' not in line:
            if '*/' in line:
                line = line[:line.find('/*')] + ' ' + line[line.find('*/') + 2:]
            else:
                line = line[:line.find('/*')]
                remove = True
        if remove:
            if '*/' in line:
                line = line[line.find('*/') + 2:]
                remove = False
            else:
                continue
        if len(line) and len(line.replace(' ', '')):
            text += line.rstrip(' ') + '\n'
    texta = text

    def resolve (f):
        if f[:2] == './':
            f = f[2:]
        if f[0] != '.' and f[0] != '/':
            f = '/'.join(file.split('/')[:-1]) + '/' + f
        elif f.startswith('../'):
            i = 0
            while f.startswith('../'):
                f = f[3:]
                i += 1
            split = '/'.join(file.split('/')[:-1 - i])
            f = split + '/' + f if len(split) else f
        return f if f[-3:] == '.js' else f + '.js'

    files = {}
    i = text.find('import ')
    while i > -1:
        if i != 0 and text[i - 1] not in ['\n', ' ']:
            text = text[i + 6:]
            i = text.find('import ')
            continue
        text = text[i:]
        i = 6
        while text[i] == ' ':
            i += 1
        if text[i] == '{':
            text = text[i:]
            k = text.find('}')
            names = text[1:k].split(',')
            names = [name.replace(' ', '') for name in names]
            names = [name for name in names if name not in ['', '{', '}']]
            text = text[k:]
            if ' from ' in text:
                i = text.find(' from ') + 6
                while text[i] == ' ':
                    i += 1
                if text[i] == "'":
                    text = text[i + 1:]
                    f = resolve(text[:text.find("'")])
                    files[f] = files[f] + names if f in files else names
        else:
            names = []
            if ' from ' in text:
                j = text.find(' ')
                k = text.find(' from ')
                while j < k:
                    while text[i] == ' ':
                        i += 1
                    name = text[i:k].split(' ')[0]
                    if name not in ['', '{', '}']:
                        names += [name]
                    i = j
                    text = text[j:]
                    if ',' in text:
                        j = text.find(',')
                    k = text.find(' from ')
                text = text[k:]
                i = 6
            while text[i] == ' ':
                i += 1
            if text[i] == "'":
                text = text[i + 1:]
                f = resolve(text[:text.find("'")])
                files[f] = files[f] + names if f in files else names
        i = text.find('import ')
    modules[file] = list(files.keys())
    if len([i for i in modules[file] if i not in imported and (i not in modules or file not in modules[i])]):
        return ''
    exporta = ['async', 'class', 'const', 'default', 'function', 'let', 'var']
    repeata = ['\n', ' ', '(', ',', '.', '[']
    text = texta
    i = text.find('export ')
    while i > -1:
        text = text[i + 7:]
        for name in exporta:
            j = text.find(name)
            if j > -1 and j < 3:
                text = text[j + len(name):]
        if '\n' in text:
            names = text[:text.find('\n')]
        split = []
        if '=' not in names or '(' in names and names.find('=') > names.find('('):
            split += [names]
        else:
            while '=' in names:
                split += [names[:names.find('=')]]
                names = names[names.find('='):]
                names = names[names.find(','):] if ',' in names else ''
        names = []
        for name in split:
            while name[0] in repeata:
                name = name[1:]
            for j in repeata:
                if j in name:
                    name = name[:name.find(j)]
            names += [name]
        files[file] = names if file not in files else files[file] + names
        i = text.find('export ')

    def replace (text, past, new):
        length = len(past)
        a = 0
        while past in text[a:]:
            a += text[a:].find(past)
            if len(text) < a + 1 + length:
                return text
            cont = False
            textb = text[a - 7:a] + new
            for name in exporta:
                if name + '_' in textb:
                    cont = True
            if cont or text[a + length] in base64 or text[a - 1] in base64 + "'.":
                a += length
                continue
            text = text[:a] + new + text[a + length:]
            a += len(new)
        return text

    text = texta
    for file in files:
        string = file.replace('.', '_').replace('/', '_')
        split = string.split('_')
        string = string[:-len(split[len(split) - 1]) - 1]
        for name in files[file]:
            text = replace(text, name, name + '_' + string)
    lines = text.split('\n')
    text = ''
    for line in lines:
        if line.startswith('export default '):
            line = line[line.find('export default ') + 15:]
        if line.startswith('export '):
            line = line[line.find('export ') + 7:]
        if len(line) and not line.startswith('import '):
            text += line + '\n'
    return text

def build (file, output):
    imported = []
    imports = [file]
    modules = {}
    texts = {}
    while len(imports):
        file = imports[0]
        if file in imported:
            imports = [i for i in imports if i != file]
        else:
            texts[file] = parse(file, imported, modules)
            imports = modules[file] + imports
            if len(texts[file]):
                imported += [file]
    text = ''
    imports = []
    for file in imported:
        if file not in imports:
            text += texts[file]
            imports += [file]
    with open(output, 'w') as file:
        file.write(text)

args = sys.argv
build(args[1], args[2])